### PR TITLE
 [PD] Support for point sections in pipe/sweep

### PR DESCRIPTION
**Pull request by AjinkyaDahale**
Similar changes to PR #5170, except uses `BRepOffsetAPI_MakePipeShell` instead.

Not ready for merge, but in good enough condition for tests. Certain bugs exist which I will describe in comments as I find them.

### CODE REVIEW

**Comment by AjinkyaDahale** [comment]
See attached files. In `lofttest-5193-1` the profile is a point and it has two more sections, with "orientation mode" set to "Fixed". This pipe succeeds. However, in `lofttest-5193-2`, with the same cross sections, but reversed (so the hexagon is the profile and the other two are added as multi-sections), the "front" face is not formed.

Additionally, with "orientation mode" set to "Standard" or "Frenet" the ends of the pipe do not end in a plane and again the face is not formed.

**Comment by AjinkyaDahale** [comment]
I think I found the cause for the bug and fixed it in c74f47cb2508e08877ec3083e788c6bfd4e885fd. @donovaly can you try this out whenever you're free. I'm sure there are a whole bunch of cases where things won't work and I would like some help by getting concrete test cases.

**Comment by donovaly** [comment]
> I think I found the cause for the bug and fixed it in [c74f47c](https://github.com/FreeCAD/FreeCAD/commit/c74f47cb2508e08877ec3083e788c6bfd4e885fd). @donovaly can you try this out whenever you're free. I'm sure there are a whole bunch of cases where things won't work and I would like some help by getting concrete test cases.

I won't have time right now. There were too many new things added to PD recently and these need to be sorted out. I am also busy in releasing FC 0.19.3. So please give me some days.

**Comment by AjinkyaDahale** [comment]
> I won't have time right now. There were too many new things added to PD recently and these need to be sorted out. I am also busy in releasing FC 0.19.3. So please give me some days.

That's OK. I don't think this is blocking things I'm planning right now.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5193">pipeline status</a> for feature branch [PR_5193](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5193). Pipeline [414384473](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/414384473) was triggered at [46e5f65](https://github.com/FreeCAD/FreeCAD/pull/5193/commits/46e5f653cc6e67aebc966ea7a42812896d56d5dc). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by donovaly** [comment]
I will find time for this tomorrow.

Can you please rebase the PR to latest master since there were a lot of changes to PD the last days and maybe you could squash the different commits to one?

What is the status? Is the PR ready to be reviewed or where do you see issues?

**Comment by AjinkyaDahale** [comment]
> I will find time for this tomorrow.

Thanks!
 
> Can you please rebase the PR to latest master since there were a lot of changes to PD the last days and maybe you could squash the different commits to one?

Will take a look maybe squash is possible.
 
> What is the status? Is the PR ready to be reviewed or where do you see issues?

I think the issues are solved last I recall.

**Comment by donovaly** [comment]
Can you please rebase you PR to current master? As it is it is based on master from Monday and I cannot compile because of merge conflicts.

File: src/Mod/PartDesign/App/FeaturePipe.cpp
Comment (donovaly) [inline src/Mod/PartDesign/App/FeaturePipe.cpp:377]: You added a strange character

File: src/Mod/PartDesign/App/FeaturePipe.cpp
Comment (donovaly) [inline src/Mod/PartDesign/App/FeaturePipe.cpp:359]: Why this change? the explicit type name helps others to understand the code. Therefore as long as the type name is there, we can keep it.

File: src/Mod/PartDesign/Gui/Command.cpp
Comment (donovaly) [inline src/Mod/PartDesign/Gui/Command.cpp:1062]: Where does the 6 comes from? Can you please add a comment?

nit-pick: our style is to have a space after commas.

**Review by donovaly — CHANGES_REQUESTED** [review]
Looks good. As I wrote you must rebase since I cannot compile.

I have only minor annotations.

File: src/Mod/PartDesign/Gui/Command.cpp
Comment (AjinkyaDahale) [reply src/Mod/PartDesign/Gui/Command.cpp:1062]: 6 is for the characters in "Vertex". A snippet copied from elsewhere in PD, which is why the space-less comma.

**Comment by AjinkyaDahale** [comment]
> Can you please rebase you PR to current master? As it is it is based on master from Monday and I cannot compile because of merge conflicts.

As far as I can tell it is already rebased on to FC master.

Screenshot at 2021-11-28 11-20-48

File: src/Mod/PartDesign/App/FeaturePipe.cpp
Comment (AjinkyaDahale) [reply src/Mod/PartDesign/App/FeaturePipe.cpp:377]: That would be my emacs again. This isn't the first time it happened and unfortunately won't be the last.

File: src/Mod/PartDesign/App/FeaturePipe.cpp
Comment (AjinkyaDahale) [reply src/Mod/PartDesign/App/FeaturePipe.cpp:359]: I think this one can stay. The for loop already says that `s` runs through every element of `shells`.

File: src/Mod/PartDesign/Gui/Command.cpp
Comment (donovaly) [reply src/Mod/PartDesign/Gui/Command.cpp]: You use the same code for lofts.

But is this check safe? What subNames could we have?

File: src/Mod/PartDesign/App/FeaturePipe.cpp
Comment (donovaly) [reply src/Mod/PartDesign/App/FeaturePipe.cpp:359]: This is nit-picking, I know but once the explicit type name is there, just keep it.

**Comment by donovaly** [comment]
> As far as I can tell it is already rebased on to FC master.

Indeed, I deleted now my local branch for your PR and recreated it and now its history shows it is based on master. It seems that the tool "GitHub Desktop" I use has some issues with stashing.

File: src/Mod/PartDesign/Gui/Command.cpp
Comment (AjinkyaDahale) [reply src/Mod/PartDesign/Gui/Command.cpp]: From https://www.cplusplus.com/reference/string/string/substr/:

> len
>    Number of characters to include in the substring (if the string is shorter, as many characters as possible are used).

So we probably don't need the length check.

Looking at that site, I see a clearer (and probably faster and safer) method https://www.cplusplus.com/reference/string/string/compare/.

As for what other subNames we can have, it includes "Edge<number>" and "Face<number>". I expect after topo-naming is merged the <number> part will be replaced by a description of where those entities are within the solid but the first part remain the same. If that's not the case, we'll have to address it in the future.

**Comment by donovaly** [comment]
I could now compile and there is only one missing feature -> for a further PR:

- Take this file:
[loftbug.zip](https://github.com/FreeCAD/FreeCAD/files/7613956/loftbug.zip)
- open there the sweep and make it a multisection
- select as first section the vertex, then the circle

result: you correctly chatched this mistake case:
FreeCAD_WJ6InnY043

Now the user knows what to do but he cannot move the order of the sections in the dialog like he can for the sections of lofts:

opSGxOh06e

File: src/Mod/PartDesign/Gui/Command.cpp
Comment (donovaly) [reply src/Mod/PartDesign/Gui/Command.cpp]: Since you also check for "Vertex", you can omit the length check I think.
Please decide.

From my point of view it is then ready to be merged.

**Comment by AjinkyaDahale** [comment]
> I could now compile and there is only one missing feature -> for a further PR:
> 
>     * Take this file:
>       [loftbug.zip](https://github.com/FreeCAD/FreeCAD/files/7613956/loftbug.zip)
> 
>     * open there the sweep and make it a multisection
> 
>     * select as first section the vertex, then the circle
> 
> 
> result: you correctly chatched this mistake case: 

So the mistake was caught by OCCT (that's the only place where the term "punctual section" is used). For some reason the algorithm requires that only the first and last added sections can be punctual sections.

File: src/Mod/PartDesign/App/FeaturePipe.cpp
Comment (AjinkyaDahale) [reply src/Mod/PartDesign/App/FeaturePipe.cpp:359]: If you insist. I'll revert this change and force push shortly.

File: src/Mod/PartDesign/Gui/Command.cpp
Comment (AjinkyaDahale) [reply src/Mod/PartDesign/Gui/Command.cpp]: So this is what I'll do:
1. In `FeaturePipe` the comparison only relates to pipe, so I'll change to using `compare`
2. In `Command.cpp` and other places where there are other occasions where this snippet is used, I'll instead make a separate PR.

Does that sound good?

Comment (donovaly) [reply src/Mod/PartDesign/Gui/Command.cpp]: Very good.

**Comment by AjinkyaDahale** [comment]
> Now the user knows what to do but he cannot move the order of the sections in the dialog like he can for the sections of lofts:

For other (i.e. not "punctual") sections, the algorithm is fairly forgiving. For example, the pipe will still work the same regardless of whether you add the rectangle first or the circle (still compiling to confirm this, but fairly sure of it). My guess is the algo guesses an appropriate position for the section along the path.

**Comment by donovaly** [comment]
I'll merge the PR now.

Concerning the multisection ordering feature, I will make a PR for this

**Comment by donovaly** [comment]
> Concerning the multisection ordering feature, I will make a PR for this

Here is the PR: https://github.com/FreeCAD/FreeCAD/pull/5207
Can you please test it?

**Comment by AjinkyaDahale** [comment]
> Concerning the multisection ordering feature, I will make a PR for this

Alright. I'll leave it to you.

**Comment by donovaly** [comment]
> Alright. I'll leave it to you.

Seems we had a mid-air collision since the same time you sent your post I sent one with the PR. Please try this out.

File: src/Mod/PartDesign/Gui/Command.cpp
Comment (AjinkyaDahale) [reply src/Mod/PartDesign/Gui/Command.cpp]: See https://github.com/FreeCAD/FreeCAD/pull/5208.

**Pull request merged**
